src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1433 @@
+            // and we never want to affect people who express their patterns thoroughly (but correctly)
+            // such as `switch { < 0 => -1, 0 => 0, > 0 => 1 }`
+            // So we're only reporting a warning for situations that syntactically look hazardous.
+            // Others are reported as a hidden diagnostic.
+            // At the moment, we're only interested in patterns in an `or` pattern with a `not` in an early case of the `or`.
+            static bool detectNotOrPattern(SyntaxNode syntax)
+            {
+start:
+                syntax = enclosingParenthesizedOrSelf(syntax);
+
+                if (isRightOfOrPattern(syntax, out var binary)
+                    && detectNotPatternOnLeftOfOrPattern(binary))
+                {
+                    return true;
+                }
+
+                Debug.Assert(syntax.Parent is not null);
+                if (isLeftOfBinaryPattern(syntax)
+                    && isRightOfOrPattern(enclosingParenthesizedOrSelf(syntax.Parent), out BinaryPatternSyntax? parentBinary)
+                    && detectNotPatternOnLeftOfOrPattern(parentBinary))
+                {
+                    return true;
+                }
+
+                // If the syntax is the whole sub-pattern, we walk up to the recursive pattern.
+                // For example: `not A or { Prop: <redundant> }`
+                if (syntax.Parent is SubpatternSyntax subpatternSyntax
+                    && subpatternSyntax.Parent is (PropertyPatternClauseSyntax or PositionalPatternClauseSyntax) and var patternClause
+                    && patternClause.Parent is RecursivePatternSyntax recursive)
+                {
+                    syntax = recursive;
+                    goto start;
+                }
+
+                // If the syntax is the whole list element pattern, we walk up to the list pattern.
+                // For example: `not A or [<redundant>, ...]`
+                if (syntax.Parent is ListPatternSyntax listPattern)
+                {
+                    syntax = listPattern;
+                    goto start;
+                }
+
+                if (syntax.Parent is SlicePatternSyntax slicePattern)
+                {
+                    syntax = slicePattern;
+                    goto start;
+                }
+
+                return false;
+            }
+
+            // Detect a `not` at top-level or inside a tree of `or` patterns
+            // Note: we don't dig into parenthesized patterns as the meaning of `not` is not problematic then
+            static bool detectNotPatternInPattern(SyntaxNode syntax)
+            {
+                if (syntax.Kind() == SyntaxKind.NotPattern)
+                {
+                    return true;
+                }
+
+                if (syntax is BinaryPatternSyntax binarySyntax && binarySyntax.Kind() == SyntaxKind.OrPattern)
+                {
+                    var stack = ArrayBuilder<BinaryPatternSyntax>.GetInstance();
```

Comment:
>var stack = ArrayBuilder<BinaryPatternSyntax>.GetInstance(); [](http://example.com/codeflow?start=20&length=60)

I think the stack is not necessary for the purpose of this method.  It looks like we could descend on the left side of `or` patterns in a loop, calling ```detectNotPatternInPattern``` for the right children on the way. And calling ```detectNotPatternInPattern``` for the left-most leaf at the very end. The order of "visiting" for children is not significant for the purpose of this helper.